Add middleware provider

MoradiDavijani opened the pull request:
It adds `osjs/middleware` provider to the core. As an example, it can be used to dynamically add items to the context menu in the file manager application.

Closes https://github.com/os-js/OS.js/issues/752.

andersevenrud commented:
I'll do a proper review of this in a little bit. But so far it looks really good! :+1: 

There's one piece of the puzzle here remaining, which is noted in https://github.com/os-js/OS.js/issues/752#issuecomment-703812143.

Since packages needs to register this middleware before the application is launched, there needs to be some kind of action that loads a script from the package without actually launching the package (application).

This could probably be done in another PR though :thinking:

MoradiDavijani commented:
Thanks 😊

Yes, I will create another PR (soon 😁) in the file manager application to make use of this middleware. And then, I think we can go for adding the middleware.js and its preload strategy in any application that needs to utilize a middleware. Is it right or we need to handle the preload strategy in this PR?

andersevenrud reviewed (changes requested):
Can you amend the commit 981d0a45575dc116309938e28a2045aa4a5f12bd message to say the following:

```
Fixed typo in clipboard type definitions
```

andersevenrud reviewed (changes requested):
Can you amend the commit 1f1548b20509e669469398b12fe214729544f7b3 message to reference the following issue I've opened about this:

https://github.com/os-js/osjs-client/issues/144

andersevenrud commented on __tests__/middleware.js:
```diff
@@ -0,0 +1,51 @@
+
+    expect(middleware[0]).toEqual(callback);
+  });
+});
```
Maybe add a test for `get` as well here ?

MoradiDavijani replied:
It would be like the test for `add`, isn't it?

andersevenrud replied:
I was thinking of having a test that loops over the callback list and checks if they all were called correctly, ex:

```
const callback = jest.fn()
// ...
expect(callback.mock.calls.length).toBe(1)
```

andersevenrud commented on src/middleware.js:
```diff
@@ -0,0 +1,102 @@
+    /**
+     * @type {MiddlewareData}
+     */
+    this.middleware = undefined;
```
This should be an empty object according to the defined type.

andersevenrud commented on src/middleware.js:
```diff
@@ -0,0 +1,102 @@
+     */
+    this.middleware = undefined;
+
+    this.clear();
```
If the default value above is changed, then this is not needed.

andersevenrud commented on src/providers/core.js:
```diff
@@ -500,6 +519,17 @@ export default class CoreServiceProvider extends ServiceProvider {
+  createMiddlewareContract() {
+    return {
+      add: (group, callback) => this.middleware.add(group, callback),
+      get: group => this.middleware.get(group),
```
Remove the trailing comma.

andersevenrud commented:
Could you amend your commit message to reference the new issue I've opened on this https://github.com/os-js/osjs-client/issues/144 ?

>  Is it right or we need to handle the preload strategy in this PR?

It's OK to commit to this PR if you want.

andersevenrud commented:
>  Is it right or we need to handle the preload strategy in this PR?

It's OK to commit to this PR if you want.

andersevenrud commented:
Looks good!

Now, I'm a bit nit-picky :joy: Would it be too much of me to ask you to squash the following commits together ?

* Added middleware provider (#144)
* Fixed some minor code styles issues
* Added a scenario in middleware tests

andersevenrud commented:
My man! I'll paste my gitter response about the loading strategy in #144 and we can take the rest from there.

andersevenrud commented:
Man, you're fast! Gotta say I'm impressed :nerd_face:

MoradiDavijani commented:
It seems that something is broken 😅 Do you know what's the problem?

andersevenrud commented on src/packages.js:
```diff
@@ -393,9 +394,13 @@ export default class Packages {
+        .map((files = [], ...iter) => ({
           type: 'application',
-          files: [],
+          files: files.map(file =>
```
Try doing `(iter.files || [])` here and remove the destructure from `.map()` to fix https://github.com/os-js/osjs-client/pull/143#issuecomment-757379309.

andersevenrud replied:
Clarification: `files: (iter.files || [])` and then remove the destructure from the `.map()` **above** :sweat_smile:

andersevenrud commented on src/packages.js:
```diff
@@ -395,8 +396,12 @@ export default class Packages {
       const append = list
         .map(iter => ({
```
I think maybe it's a good idea to split out this mapping into a utility function in `src/utils/packages.js`.

andersevenrud replied:
Doing this will fix the codeclimate complaint as well :)

andersevenrud commented on src/packages.js:
```diff
@@ -451,6 +456,33 @@ export default class Packages {
+   * @param {string} type Type of files to return
+   * @return {string[]}
+   */
+  _getPreloadFileNames(files, type) {
```
I think maybe it's a good idea to split out this method into a utility function in src/utils/packages.js instead.

MoradiDavijani replied:
what's your suggestion for their names? 😅

andersevenrud replied:
The one from `addPackages` could be named something like `constructManifest` and the `_getPreloadFileNames`  one maybe `fileNamesFromPreload` ?!

andersevenrud replied:
Maybe not `constructManifest`, but `createManifestFromArray` to be a bit more explicit :nerd_face: 

Maybe you have some better suggestions ?

MoradiDavijani replied:
It's perfect 😁

MoradiDavijani replied:
Maybe `fileNamesFromPreload` is not precise. We want to filter `files` by their `type` and return their `filename`s. Something like `getFileNamesByType`?

andersevenrud replied:
I generally only use verbs like "get" or "set" on class members. But in this case I can't really think of anything outside maybe `preloadFilenamesByType` :thinking:

MoradiDavijani replied:
`preloadFilenamesByType` is like it's preloading something. How about `filesToFileNames`?

andersevenrud replied:
Why not just embrace functions a bit more and introduce a higher-order-function for the filtering? :blush: 

```
.map(filterPreloadFilesByType('type'))
.map(preloadFilesToFilename)
```

MoradiDavijani replied:
It will be like this:

```javascript
filenames = preloadFilesToFilename(filterPreloadFilesByType('type'))
```

Also, since it contains both "preload" and "background" types now, isn't it better to call them `files` instead of `preloadFiles`? So:

```javascript
filenames = filesToFilename(filterFilesByType('type'))
```

BTW, IMO it's better to be a single function because it's being used in 3 places.

andersevenrud replied:
> Also, since it contains both "preload" and "background" types now, isn't it better to call them files instead of preloadFiles? So:

Yeah. That'll probably do fine... it's not like this conflicts with any other exported functions, and it is "namespaced" with the filename.

But if you want to be really explicit I'm open to calling it maybe `metadataFiles`.

> BTW, IMO it's better to be a single function because it's being used in 3 places.

Okay. Let's see how this looks when you change the function names :blush:

MoradiDavijani replied:
Naming is really the hardest thing in programming😂 Is it fine?

```javascript
filenames = metadataFilesToFilenames(filterMetadataFilesByType('type'))
```

andersevenrud replied:
> Naming is really the hardest thing in programming

Yeah... I've wasted so many hours (and energy) of my life by naming everything "perfectly". :notes: But in the end it doesn't even matter (insert meme here)! :notes: 

> Is it fine?

Yeah. This looks great!

andersevenrud reviewed (changes requested):
The things I've commented in this request is about methods that could be separated from this class as they do not directly rely on it.

How do you feel about this ?

andersevenrud commented on src/utils/packages.js:
```diff
@@ -62,3 +62,23 @@ export const createPackageAvailabilityCheck = (core) => {
   return metadata => checks.every(fn => fn(metadata));
 };
+
+export const createManifestFromArray = list => {
```
Remove the `return` statement and collapse into

```
export const createManifestFromArray = list => list
  /* ... */
```

andersevenrud commented on src/packages.js:
```diff
@@ -53,7 +58,7 @@ import {createPackageAvailabilityCheck} from './utils/packages';
  * @property {string} [server] Server script filename
  * @property {string[]} [groups] Only available for users in this group
- * @property {string[]} [files] Files to preload
+ * @property {Object[]|string[]} [files] Files to preload
```
I think the only thing left here is to update the TS type definition :blush:

andersevenrud replied:
I actually have a script to generate this from the esdoc/jsdoc... I need to add that stuff into the repo.

andersevenrud commented on src/packages.js:
```diff
@@ -240,11 +246,12 @@ export default class Packages {
+    const preloads = metadataFilesToFilenames(
+      filterMetadataFilesByType(metadata.files, 'preload')
+    ).map(f => this.core.url(f, {}, ({
+      type: metadata.type,
```
Hm. Looks like I did this.... but the `type: metadata.type` does't do anything here since it assigns `metadata` on the next line :joy: Not sure what I thought, but it would probably be one of those:

* `type: metadata.type || 'theme'`
* `type: 'theme'`

:joy:

MoradiDavijani replied:
😂😂

Which one is correct? 😁

MoradiDavijani replied:
It's should not be the first one because `metadata.type` is one of `['theme', 'icons', 'sounds']`.

andersevenrud replied:
The second one is the "correct" one, but I'm not sure what I intended when I wrote this, hehe.

andersevenrud commented:
Codeclimate brings up a good point in this latest report. These blocks could be replaced with a method in the `Packages` class that accepts the "file type" and "package type" as parameters.

andersevenrud commented:
Good stuff!

Sorry if I dragged this process out a bit, but I'm a bit of a perfectionist when it comes to certain things :sweat_smile: I really enjoyed this though! Very fast response, good insight and feedback, and you really seem to know your stuff. Love it! hehe

MoradiDavijani commented:
I think it was easier and faster for you to do it yourself, so thanks for letting me working on it. I really enjoyed working on this issue with you ✌️

andersevenrud commented:
I'm going to test this out myself tomorrow (well... later today, it's 2AM and going to sleep soon) to make sure this is ready for a merge.

MoradiDavijani commented:
@andersevenrud I found a bug (the url was not correct) while testing it with filemanager application. It's fixed now.

andersevenrud commented on index.d.ts:
```diff
@@ -1751,7 +1751,7 @@ export type PackageMetadata = {
 	 * Files to preload
 	 */
-	files?: string[];
+	files?: (object | string)[];
```
I don't think this is valid :thinking: 

```
files?: object[] | string[];
```

MoradiDavijani replied:
I think `files?: object[] | string[];` means it should be an array of objects OR an array of strings. But, `(object | string)[]` means it's an array of objects or strings. I'm not an expert in Typescript but let me know if I'm wrong.

andersevenrud replied:
Ah, yes. I think I prefer the following syntax when defining a union type like this `Array<object | string>` :thinking:

andersevenrud commented:
Where in the code was this ? I can't see the actual fix because of the force push :blush:

MoradiDavijani commented on src/packages.js:
```diff
@@ -452,6 +463,22 @@ export default class Packages {
+  _preloadBackgroundFiles(list) {
+    const backgroundFiles = list.reduce((filenames, iter) => [
+      ...filenames,
+      ...this._getPreloads(iter, 'background', 'apps')
```
@andersevenrud I changed this line. It was like this:
```js
...filenames,
...metadataFilesToFilenames(
  filterMetadataFilesByType(metadata.files, fileType)
)
```

andersevenrud replied:
Alrighty.

andersevenrud commented:
I think something went wrong with your amend because the commit message changed ?!

It's now `Added support for middleware in edit menu`

MoradiDavijani commented:
@andersevenrud Sorry for the commit message it was because I was working on filemanager at the same time.

I added a message on the code I've changed, did you see that?

andersevenrud commented:
If you mean the conversation I started (https://github.com/os-js/osjs-client/pull/143#issuecomment-761857366), then yeah. I answered and resolved that.

andersevenrud commented:
PR in FileManager: https://github.com/os-js/osjs-filemanager-application/pull/31

MoradiDavijani commented:
I'm getting the following new warnings with these changes:

![Screen Shot 1399-10-28 at 23](https://user-images.githubusercontent.com/12628677/104854892-88909600-591e-11eb-9c41-1383911f4c6a.jpg)

Do you have any idea what's the issue?

andersevenrud commented:
These changes does not affect anything related to that, and those messages are normal :)

They go away when you activate the desktop icons and put some stuff in there.
